### PR TITLE
feat: integrates database container

### LINKED ISSUE #93: Integrates mongodb in server package
## Expected Behavior
Mongodb via docker

## Actual Behavior
No backend

### PR DESCRIPTION
**Pull request by guilhermevrs**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: Closes #93 

## What is the new behavior?

- run `yarn workspace @cards/server start:db` to launch the containers
- access localhost:8081 for mongodb GUI
- On preinstall, check is run to see if you have all the prerequisites

## Does this introduce a breaking change?

- [ ] Yes
- [X] No

## A picture tells a thousand words

image

### CODE REVIEW

File: packages/server/db/init-mongo.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    user: 'appadmin',
+    pwd: 'appadminpassword',
+    roles: ['readWrite']
+})
+
+cardzdb.createCollection('users');
+cardzdb.createCollection('games');
+cardzdb.createCollection('logs');
+
+print('END #################################################################');
```
Comment (tib-tib) [inline packages/server/db/init-mongo.js:15]: is this file automatically executed when starting mongo?

**Comment by tib-tib** [comment]
In general I prefer having the infrastructure separated from the server. Maybe you could create the same structure but in an `infrastructure` package? Also I don't use `mongo-express`, I usually go with https://robomongo.org/ but I don't mind if you keep it, I just won't use it 🙃

File: packages/server/db/init-mongo.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+print('Start #################################################################');
+
+cardzdb = db.getSiblingDB('cardzdb');
+
+cardzdb.createUser({
+    user: 'appadmin',
+    pwd: 'appadminpassword',
```
Comment (tib-tib) [inline packages/server/db/init-mongo.js:7]: even if it is used just for local I prefer to use environment variables as much as possible and to no have hardcoded login information in the code 😉

File: packages/server/scripts/precheck.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+results.forEach((result) => {
+    if (result.error) {
+        console.log(result.message);
+        error = true;
+    }
+});
+
+if (error) {
+    throw new Error('Failed precheck');
+}
```
Comment (tib-tib) [inline packages/server/scripts/precheck.js:25]: To be honest this file looks a bit overkill to me. If you specify in the Readme that `docker-compose` is a prerequisite it should be enough.

File: packages/server/db/init-mongo.js (hunk 1 above)
Comment (guilhermevrs) [reply packages/server/db/init-mongo.js:15]: Correct!
This is to initialise the container if it's the first time it's being built.

File: packages/server/db/init-mongo.js (hunk 2 above)
Comment (guilhermevrs) [reply packages/server/db/init-mongo.js:7]: Indeed, even if it's local, I guess it's better to keep it in the .env (like the other credentials that are already in it)

File: packages/server/scripts/precheck.js (hunk 3 above)
Comment (guilhermevrs) [reply packages/server/scripts/precheck.js:25]: Alright, although it's not only checking for the docker-compose but also ensuring you have a .env file in it.
This kind of error could be tricky and I feel that only marking that on the README is a bit lazy, don't you think?

I could, though, make a default value for the variables, so at least we won't have errors.
What do you think?

File: packages/server/db/init-mongo.js (hunk 2 above)
Comment (guilhermevrs) [reply packages/server/db/init-mongo.js:7]: Actually, I removed the creation of user - at least until we really start using the db, we won't be needing anything different from the admin one

File: packages/server/scripts/precheck.js (hunk 3 above)
Comment (guilhermevrs) [reply packages/server/scripts/precheck.js:25]: I did the above and changed the docs, let me know if it's cool for you

**Review by tib-tib — APPROVED** [review]

**Pull request merged**